recipes/foxglove-schemas-protobuf/all/conanfile.py

```diff
@@ -0,0 +1,101 @@
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout
+from conan.tools.build import check_min_cppstd
+from conan.tools.files import get, copy, apply_conandata_patches, export_conandata_patches
+from conan.tools.microsoft import check_min_vs
+from conan.tools.scm import Version
+import os
+
+
+required_conan_version = ">=1.53.0"
+
+
+class FoxgloveSchemasProtobufConan(ConanFile):
+    name = "foxglove-schemas-protobuf"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/foxglove/schemas"
+    description = "Protobuf schemas for Foxglove"
+    license = "MIT"
+    topics = ("foxglove", "protobuf", "schemas")
+
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+    }
+
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "CMakeDeps"
+    package_type = "library"
+
+    @property
+    def _min_cppstd(self):
+        return 17
+
+    @property
+    def _compilers_minimum_version(self):
+        return {
+            "msvc": "191",
+            "gcc": "9",
+            "clang": "9",
+            "apple-clang": "12",
+        }
+
+    def export_sources(self):
+        copy(self, "CMakeLists.txt", src=self.recipe_folder, dst=(self.export_sources_folder + "/src"))
+        export_conandata_patches(self)
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def validate(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            check_min_cppstd(self, self._min_cppstd)
+        check_min_vs(self, self._compilers_minimum_version["msvc"])
+        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
+        if minimum_version and Version(self.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                f"{self.ref} requires C++{self._min_cppstd}, which your compiler does not support."
+            )
+        if self.settings.os == "Windows" and self.options.shared:
+            raise ConanInvalidConfiguration("Windows shared builds are not supported yet.")
+
+    def build_requirements(self):
+        self.tool_requires("protobuf/3.21.9")
+
+    def requirements(self):
+        self.requires("protobuf/3.21.9", transitive_headers=True, transitive_libs=True)
```

Comment:
Protobuf is very strict when generating header, pushing the specific version was used to create .hh and .cc files. I should validate if this recipe needs or not to depends on full_package_mode for protobuf. Otherwise, consumer could have a compiler error when using a different Protobuf version.

Reply:
@kevswims correct, like I said, protobuf is very strict when generating headers and sources, so this package is not a regular library, but a product from proto files, it's another thing, the regular rules of compatibility may not fit in case a consumer uses another version of protobuf, but the headers are not compatible due the explicit define made by protobuf.

Reply:
@uilianries I understand that part. Is there a specific thing that needs to be changed in the recipe to enforce that? I am not familiar with the full_package_mode settings and it does not even appear to be an option in conan 2.0.

Reply:
@kevswims let me confirm it today. But in case you would need to do it:

```python
def package_id(self):
    self.info.requires["protobuf"].patch_mode()
```

https://docs.conan.io/2/reference/conanfile/methods/requirements.html#package-id-mode

https://docs.conan.io/2/reference/conanfile/attributes.html#package-id-embed-non-embed-unknown-mode

Reply:
All fine, protobuf is minor compatible. I checked locally right, by building different applications, using different patch version of Protobuf and adding them a very same application. No errors.

Plus, it's documented about its backward compatibility with minor versions: https://github.com/protocolbuffers/protobuf/blob/532d2a24800890890d764e2de8fc2716eb3213e9/java/README.md?plain=1#L142